hudi-common/src/test/java/org/apache/hudi/common/table/read/TestEngineBasedMerger.java

```diff
@@ -0,0 +1,352 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.common.table.read;
+
+import org.apache.hudi.common.config.RecordMergeMode;
+import org.apache.hudi.common.config.TypedProperties;
+import org.apache.hudi.common.engine.HoodieReaderContext;
+import org.apache.hudi.common.model.HoodieKey;
+import org.apache.hudi.common.model.HoodieRecord;
+import org.apache.hudi.common.model.HoodieRecordMerger;
+import org.apache.hudi.common.model.OverwriteWithLatestAvroPayload;
+import org.apache.hudi.common.table.HoodieTableConfig;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.common.util.collection.Pair;
+
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.generic.IndexedRecord;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.junit.jupiter.params.provider.ValueSource;
+import org.mockito.ArgumentCaptor;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Properties;
+import java.util.stream.Stream;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertSame;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.RETURNS_DEEP_STUBS;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+class TestEngineBasedMerger {
+  private static final BufferedRecord<TestRecord> T1 = new BufferedRecord<>("key", 1, new TestRecord(), 2, false);
+  private static final BufferedRecord<TestRecord> T2 = new BufferedRecord<>("key", 2, new TestRecord(), 2, false);
+  private static final BufferedRecord<TestRecord> T3 = new BufferedRecord<>("key", 3, new TestRecord(), 2, false);
+  private static final BufferedRecord<TestRecord> HARD_DELETE = new BufferedRecord<>("key", 0, new TestRecord(), 2, true);
+  private static final BufferedRecord<TestRecord> T2_SOFT_DELETE = new BufferedRecord<>("key", 2, new TestRecord(), 2, true);
+
+  private final HoodieReaderContext<TestRecord> readerContext = mock(HoodieReaderContext.class, RETURNS_DEEP_STUBS);
+  private final TypedProperties props = mock(TypedProperties.class);
+  private final Schema readerSchema = mock(Schema.class);
+
+  private static Stream<Arguments> commitTimeOrdering() {
+    return Stream.of(
+        // Validate commit time does not impact the ordering
```

Comment:
I think this is correct?:

// Validate event time does not impact the ordering